File: typescript/two_sum.ts

function getTwoSum(nums: number[], target: number): [number, number] | null {

    const seen: Record<number, number> = {};

    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];

        if (seen.hasOwnProperty(complement)) {
            return [seen[complement], i];
        }

        seen[nums[i]] = i;
    }

    return null; // Like Go's 'nil' return when no match
}

function main() {
    const nums = [1, 2, 3, 4];
    const target = 7;

    const result = getTwoSum(nums, target);

    if (result === null) {
        console.log("Error: No match found.");
    } else {
        console.log(`Response: [${result[0]}, ${result[1]}]`);
    }
}

main();

File: cpp/two_sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> twoSum(const vector<int>& nums, int target) {
    unordered_map<int, int> numToIndex;  // value -> index
    for (int i = 0; i < nums.size(); ++i) {
        int complement = target - nums[i];
        if (numToIndex.count(complement)) {
            return { numToIndex[complement], i };
        }
        numToIndex[nums[i]] = i;
    }
    return {};  // no solution
}

int main() {
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;

    vector<int> result = twoSum(nums, target);
    if (!result.empty()) {
        cout << "Indices: " << result[0] << ", " << result[1] << endl;
    } else {
        cout << "No two sum solution found." << endl;
    }

    return 0;
}

